docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+# [RFC] Conjure Endpoint Errors
+
+**Date:** 2020/03/19
+
+**Stakeholders:**
+* *Anyone that produces or consumes conjure-generated code.*
+
+* * *
+
+## **Context**
+
+Currently, errors can be defined in Conjure type definitions. These errors can then be thrown by services, and if a client 
+knows about them, they can be handled during runtime by clients. The exact handling of this mechanism may vary by
+the implementation (for example, in Java these are `ServicException`s, which become `RemoteException`s which can be caught
+by the client).
+
+There are multiple issues with this. As a consumer, you only know that an endpoint may throw an error by reading the 
+documentation or the implementation itself. As a producer, you are limited in the expressivity of what you can say 
+with your exception, as you have no guarantee that a consumer will handle your exception (or even read your documentation).
+
+This RFC introduces *declared errors* on Conjure endpoints. These errors represent real outcomes of a call to an endpoint, 
+and should be handled by consumers.
+
+## Existing Documentation
+
+### PRs for example implementation (in Java, using checked exceptions)
+
+* Root PR (Conjure API changes): https://github.com/palantir/conjure/pull/816
+* Conjure Java API Changes: https://github.com/palantir/conjure-java-runtime-api/pull/634
+* Conjure Java & Dialogue implementation: https://github.com/palantir/conjure-java/pull/1275
+
+## Example Conjure and Java implementations
+
+**Conjure API**
+
+```
+types:
+  definitions:
+    default-package: com.palantir.my.api
+    errors:
+      BlueMoon:
+        namespace: MyConjureService
+        code: CUSTOM_SERVER
+
+services:
+  MyConjureService:
+    name: An example service.
+    package: com.palantir.my.api
+    default-auth: header
+    endpoints:
+      getSomeResult:
+        http: GET /
+        returns: list<rid>
+        errors:
+          - BlueMoon
+        docs: Returns some result, but once in a blue moon throws an error
+```
+
+
+**Undertow Service**
+
+```
+@Generated("com.palantir.conjure.java.services.UndertowServiceInterfaceGenerator")
+public interface MyConjureService {
+    /**
+     * Returns some result, but once in a blue moon throws an error
+     * @apiNote {@code GET /}
+     */
+    List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+            throws MyConjureServiceErrors.BlueMoonServiceException;
+```
+
+**Undertow Resource**
+
+```
+public final class MyConjureResource implements MyConjureService {
+
+   ...
+
+   @Override
+   public List<ResourceIdentifier> getSomeResult(AuthHeader authHeader)
+          throws BlueMoonServiceException {
+      throw MyConjureServiceErrors.throwIfBlueMoon(isBlueMoon);
+   }
+```
+
+### Example Dialogue Client Implementation via Exceptions
```

Comment:
Its worth noting that checked exceptions would not work with async clients. Have we thought about how we would implement declared errors in this case? It just seems like a hard drop in safety if you move to async clients

Reply:
> Its worth noting that checked exceptions would not work with async clients.

I'd expect the return type to become more complex than the existing ListenableFuture. The future is already a form of special `Result` object. Async clients are already used much less broadly than other clients and much easier to misuse. That's not to say we should make it easier to misuse, but I wouldn't use it as an argument against adding safety to most common use cases.